s3/src/main/java/org/togglz/s3/S3StateRepository.java

```diff
@@ -0,0 +1,118 @@
+package org.togglz.s3;
+
+import org.togglz.core.Feature;
+import org.togglz.core.repository.FeatureState;
+import org.togglz.core.repository.StateRepository;
+import org.togglz.core.util.FeatureStateStorageWrapper;
+
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.S3Object;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * A state repository that uses Amazon S3.
+ * <p>
+ * The repository is configured using the {@link Builder}
+ * <p>
+ * You must already have a bucket provisioned before you create this repository.
+ * 
+ * @author Mark Richardson
+ * @date 12/27/16
+ */
+public class S3StateRepository implements StateRepository {
+
+    // http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
+    private static final String ERR_NO_SUCH_KEY = "NoSuchKey";
+
+    private ObjectMapper objectMapper = new ObjectMapper();
+
+    protected final AmazonS3Client client;
+    protected final String bucketName;
+    protected final String keyPrefix;
+
+    private S3StateRepository(Builder builder) {
+        this.client = builder.client;
+        this.bucketName = builder.bucketName;
+        this.keyPrefix = builder.keyPrefix;
+    }
+
+    @Override
+    public FeatureState getFeatureState(Feature feature) {
+        try (S3Object object = client.getObject(bucketName, keyPrefix + feature.name())) {
+            if (object != null) {
+                FeatureStateStorageWrapper wrapper = objectMapper.reader().forType(FeatureStateStorageWrapper.class).readValue(object.getObjectContent());
+                return FeatureStateStorageWrapper.featureStateForWrapper(feature, wrapper);
+            }
+        } catch (AmazonS3Exception ae) {
+            if (ERR_NO_SUCH_KEY.equals(ae.getErrorCode())) {
+                return null;
+            }
+            throw ae;
+        } catch (Exception e) {

```

Comment:
Sorry, my comment was a bit unclear. My point was that catching a `java.lang.Exception` is considered bad practice. Why did you add that catch block which basically catches EVERYTHING (include `NullPointerException` for example). Does the Amazon client throw more checked exception? In this case just catch these specific types. I don't see any value in catching "everything" here. So my votes goes for removing the block that catches `java.lang.Exception`.